src/Microsoft.Diagnostics.Monitoring/DiagnosticsEventPipeProcessor.cs

```diff
@@ -0,0 +1,320 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Parsers.FrameworkEventSource;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring
+{
+    public enum PipeMode
+    {
+        Logs = 1,
+        Metrics,
+    }
+
+    public class DiagnosticsEventPipeProcessor : IAsyncDisposable
+    {
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly IEnumerable<IMetricsLogger> _metricLoggers;
+        private readonly PipeMode _mode;
+
+        //These values don't change so we compute them only once
+        private readonly List<string> _dimValues;
+
+        public const string NamespaceName = "Namespace";
+        public const string NodeName = "Node";
+        private static readonly List<string> DimNames = new List<string> { NamespaceName, NodeName };
+
+        public DiagnosticsEventPipeProcessor(ContextConfiguration contextConfig, PipeMode mode,
+            ILoggerFactory loggerFactory,
+            IEnumerable<IMetricsLogger> metricLoggers)
+        {
+            _dimValues = new List<string> { contextConfig.Namespace, contextConfig.Node };
+            _metricLoggers = metricLoggers;
+            _mode = mode;
+            _loggerFactory = loggerFactory;
+        }
+
+        public async Task Process(int pid, int duration, CancellationToken token)
+        {
+            await Task.Factory.StartNew(async () =>
```

Comment:
This call to Task.Factory.StartNew is returning a Task<Task> type. I think the intention is that the inner task should be awaited before the Process method returns, rather than waiting on the Task that is holding the Task from the async annonymous method. I believe you need to either call Unwrap on the result of Task.Factory.StartNew before awaiting or await the result of the await.

Reply:
Yeah I forgot about this detail of StartNew.